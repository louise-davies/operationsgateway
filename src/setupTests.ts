// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import crypto from 'crypto';
// TODO JOEL: Is this still needed? - Also in vite config
// need to mock <canvas> for plotting
import 'vitest-canvas-mock';
// TODO JOEL: Replace with vitest-fail-on-console
import failOnConsole from 'jest-fail-on-console';
import { TextEncoder } from 'util';
import { server } from './mocks/server';

// TODO JOEL: Is this still needed? - same for failOnConsole below
global.TextEncoder = TextEncoder;

failOnConsole();

vi.setConfig({ testTimeout: 15000 });

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

// TODO JOEL: Check if a all of the below is still needed

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {},
  });
}

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
vi.doMock('react', async () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = await vi.importActual('react');
  }
  return mockActualReact;
});

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.revokeObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {
      // no-op
    },
  });
}

// jest doesn't implement web crypto so set up nodejs crypto as a default
Object.defineProperty(global, 'crypto', {
  value: Object.setPrototypeOf({ subtle: crypto.subtle }, crypto),
});
