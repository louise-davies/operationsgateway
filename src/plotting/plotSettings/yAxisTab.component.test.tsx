import React from 'react';
import { fireEvent, render, screen, within } from '@testing-library/react';
import type { RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import YAxisTab, { ColourGenerator } from './yAxisTab.component';
import type { YAxisTabProps } from './yAxisTab.component';
import { testChannels } from '../../setupTests';
import { FullScalarChannelMetadata } from '../../app.types';

describe('y-axis tab', () => {
  let props: YAxisTabProps;
  let user;
  const changeYAxesScale = jest.fn();
  const changeSelectedPlotChannels = jest.fn();
  const changeYMinimum = jest.fn();
  const changeYMaximum = jest.fn();

  const createView = (): RenderResult => {
    return render(<YAxisTab {...props} />);
  };

  beforeEach(() => {
    props = {
      selectedRecordTableChannels: [
        {
          systemName: 'timestamp',
          channel_dtype: 'scalar',
          userFriendlyName: 'Time',
        },
      ],
      allChannels: testChannels as FullScalarChannelMetadata[],
      selectedPlotChannels: [],
      changeSelectedPlotChannels,
      changeYMinimum,
      changeYMaximum,
      YAxesScale: 'linear',
      changeYAxesScale,
    };

    user = userEvent.setup({ delay: null });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { asFragment } = createView();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Y scale radio buttons and calls changeYAxesScale on click', async () => {
    createView();

    const radioGroup = screen.getByRole('radiogroup', { name: 'Scale' });
    expect(
      within(radioGroup).getByRole('radio', {
        name: 'Linear',
      })
    ).toBeChecked();

    await user.click(screen.getByRole('radio', { name: 'Log' }));

    expect(changeYAxesScale).toHaveBeenCalledWith('logarithmic');
  });

  it('allows user to add channels on the y-axis (keyboard only)', async () => {
    createView();

    const autocomplete = screen.getByRole('autocomplete');
    const input = within(autocomplete).getByRole('combobox');

    await user.type(input, 'test_');
    autocomplete.focus();
    fireEvent.keyDown(autocomplete, { key: 'ArrowDown' });
    fireEvent.keyDown(autocomplete, { key: 'Enter' });

    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: expect.anything(),
          lineStyle: 'solid',
        },
      },
    ]);
  });

  it('allows user to add channels on the y-axis (mouse and keyboard)', async () => {
    createView();

    const autocomplete = screen.getByRole('autocomplete');
    const input = within(autocomplete).getByRole('combobox');

    await user.type(input, 'test_');
    await user.click(screen.getByText('test_1'));

    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: expect.anything(),
          lineStyle: 'solid',
        },
      },
    ]);
  });

  it('populates the displayed table channels dropdown and adds selection to the y-axis', async () => {
    props.selectedRecordTableChannels = [
      {
        systemName: 'test_1',
        channel_dtype: 'scalar',
      },
    ];
    createView();

    const select = screen.getByTestId('select displayed table channels');
    fireEvent.change(select, { target: { value: 'test_1' } });

    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: expect.anything(),
          lineStyle: 'solid',
        },
      },
    ]);
  });

  it('only populates the displayed table channels dropdown with options not already selected', async () => {
    props.selectedRecordTableChannels = [
      {
        systemName: 'test_1',
        channel_dtype: 'scalar',
      },
    ];
    props.selectedPlotChannels = [
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ];
    createView();

    const select = screen.getByTestId('select displayed table channels');
    fireEvent.change(select, { target: { value: 'test_1' } });

    expect(changeSelectedPlotChannels).not.toHaveBeenCalled();
  });

  it('allows user to toggle visibility of a selected channel off', async () => {
    props.selectedPlotChannels = [
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Toggle test_1 visibility off'));
    expect(changeSelectedPlotChannels).toHaveBeenLastCalledWith([
      {
        name: 'test_1',
        options: {
          visible: false,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ]);
  });

  it('allows user to toggle visibility of a selected channel on', async () => {
    props.selectedPlotChannels = [
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: false,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Toggle test_2 visibility on'));
    expect(changeSelectedPlotChannels).toHaveBeenLastCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ]);
  });

  it('allows user to change plot colour of a channel', async () => {
    props.selectedPlotChannels = [
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Pick test_2 colour'));
    await user.click(screen.getByLabelText('Color'));

    expect(changeSelectedPlotChannels).toHaveBeenLastCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: expect.anything(),
          lineStyle: 'solid',
        },
      },
    ]);
  });

  it('allows user to change line style of a channel', async () => {
    props.selectedPlotChannels = [
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Change test_2 line style'));
    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'dashed',
        },
      },
    ]);
    await user.click(screen.getByLabelText('Change test_2 line style'));
    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'dotted',
        },
      },
    ]);
    await user.click(screen.getByLabelText('Change test_2 line style'));
    expect(changeSelectedPlotChannels).toHaveBeenCalledWith([
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ]);
  });

  it('removes channel from display when we click Close on its label', async () => {
    props.selectedPlotChannels = [
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Remove test_1 from plot'));
    expect(changeSelectedPlotChannels).toHaveBeenLastCalledWith([
      {
        name: 'test_2',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ]);
    expect(changeYAxesScale).not.toHaveBeenCalled();
  });

  it('removes channel from display when we click Close on its label and resets y-axis scale to linear if no selected channels remain', async () => {
    props.selectedPlotChannels = [
      {
        name: 'test_1',
        options: {
          visible: true,
          colour: '#ffffff',
          lineStyle: 'solid',
        },
      },
    ];
    createView();

    await user.click(screen.getByLabelText('Remove test_1 from plot'));
    expect(changeSelectedPlotChannels).toHaveBeenLastCalledWith([]);
    expect(changeYAxesScale).toHaveBeenCalledWith('linear');
  });

  describe('min and max fields', () => {
    it('lets user change the min field and calls relevant onchange method', async () => {
      createView();

      const minField = screen.getByLabelText('Min');
      await user.type(minField, '1');
      expect(changeYMinimum).toHaveBeenCalledWith(1);
    });

    it('lets user change the max field and calls relevant onchange method', async () => {
      createView();

      const maxField = screen.getByLabelText('Max');
      await user.type(maxField, '1');
      expect(changeYMaximum).toHaveBeenCalledWith(1);
    });

    it('displays helper text when min and max fields contain an invalid range', async () => {
      createView();

      const minField = screen.getByLabelText('Min');
      const maxField = screen.getByLabelText('Max');
      await user.type(minField, '2');
      await user.type(maxField, '1');

      // Check the helper text displays
      screen.getAllByText('Invalid range');

      // One for each input box
      expect(screen.getAllByText('Invalid range').length).toEqual(2);
    });
  });
});

describe('ColourGenerator', () => {
  let colourGenerator;
  const colourOrder = [
    '#008000', // dark green
    '#0000ff', // dark blue
    '#ff00ff', // pink
    '#00ffff', // light blue
    '#008080', // teal
    '#800000', // deep red
    '#00ff00', // light green
    '#000080', // navy blue
    '#7f8000', // brown-ish yellow?
    '#80007f', // indigo
  ];

  beforeEach(() => {
    colourGenerator = new ColourGenerator();
  });

  describe('getting next colour', () => {
    it('returns the next colour in the remaining colours list', () => {
      const colour = colourGenerator.nextColour();
      expect(colour).toEqual(colourOrder[0]);
    });

    it('returns a random colour if the list of remaining colours is empty', () => {
      colourOrder.forEach(() => {
        colourGenerator.nextColour();
      });
      const colour = colourGenerator.nextColour();
      expect(colourOrder.includes(colour)).toBeFalsy();
    });
  });

  describe('removing a colour', () => {
    it('1 colour selected', () => {
      const firstGenerated = colourGenerator.nextColour();

      colourGenerator.removeColour(firstGenerated);

      const secondGenerated = colourGenerator.nextColour();
      expect(firstGenerated).toEqual(secondGenerated);
    });

    describe('2 colours selected', () => {
      it('removes first', () => {
        const firstGenerated = colourGenerator.nextColour();
        colourGenerator.nextColour();

        colourGenerator.removeColour(firstGenerated);

        const thirdGenerated = colourGenerator.nextColour();
        expect(thirdGenerated).toEqual(firstGenerated);
      });

      it('removes second', () => {
        colourGenerator.nextColour();
        const secondGenerated = colourGenerator.nextColour();

        colourGenerator.removeColour(secondGenerated);

        const thirdGenerated = colourGenerator.nextColour();
        expect(thirdGenerated).toEqual(secondGenerated);
      });
    });

    describe('3 colours selected', () => {
      it('removes first', () => {
        const firstGenerated = colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();

        colourGenerator.removeColour(firstGenerated);

        const fourthGenerated = colourGenerator.nextColour();
        expect(fourthGenerated).toEqual(firstGenerated);
      });

      it('removes middle', () => {
        colourGenerator.nextColour();
        const secondGenerated = colourGenerator.nextColour();
        colourGenerator.nextColour();

        colourGenerator.removeColour(secondGenerated);

        const fourthGenerated = colourGenerator.nextColour();
        expect(fourthGenerated).toEqual(secondGenerated);
      });

      it('removes last', () => {
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        const thirdGenerated = colourGenerator.nextColour();

        colourGenerator.removeColour(thirdGenerated);

        const fourthGenerated = colourGenerator.nextColour();
        expect(fourthGenerated).toEqual(thirdGenerated);
      });
    });

    describe('10 colours selected', () => {
      it('removes first', () => {
        const firstGenerated = colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();

        colourGenerator.removeColour(firstGenerated);

        const fourthGenerated = colourGenerator.nextColour();
        expect(fourthGenerated).toEqual(firstGenerated);
      });

      it('removes some middle ones', () => {
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        const thirdGenerated = colourGenerator.nextColour();
        colourGenerator.nextColour();
        const fifthGenerated = colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        const ninthGenerated = colourGenerator.nextColour();
        colourGenerator.nextColour();

        // Remove third, then ninth, then fifth
        colourGenerator.removeColour(thirdGenerated);
        colourGenerator.removeColour(ninthGenerated);
        colourGenerator.removeColour(fifthGenerated);

        const eleventhGenerated = colourGenerator.nextColour();
        const twelfthGenerated = colourGenerator.nextColour();
        const thirteenthGenerated = colourGenerator.nextColour();
        expect(eleventhGenerated).toEqual(thirdGenerated);
        expect(twelfthGenerated).toEqual(fifthGenerated);
        expect(thirteenthGenerated).toEqual(ninthGenerated);
      });

      it('removes last', () => {
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        const tenthGenerated = colourGenerator.nextColour();

        colourGenerator.removeColour(tenthGenerated);

        const eleventhGenerated = colourGenerator.nextColour();
        expect(eleventhGenerated).toEqual(tenthGenerated);
      });
    });

    describe('11 colours selected', () => {
      it('removes last', () => {
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        colourGenerator.nextColour();
        const eleventhGenerated = colourGenerator.nextColour();

        colourGenerator.removeColour(eleventhGenerated);

        // No more colours in remaining colours list, so we expect a random colour
        const twelfthGenerated = colourGenerator.nextColour();
        expect(twelfthGenerated).not.toEqual(eleventhGenerated);
      });
    });
  });
});
