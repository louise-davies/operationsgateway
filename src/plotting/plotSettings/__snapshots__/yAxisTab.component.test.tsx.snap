// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`y-axis tab renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1w7ngrt-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-1 css-v8g4am-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
            style="font-size: 12px;"
          >
            Min
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            style="font-size: 12px;"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Min
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
            style="font-size: 12px;"
          >
            Max
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            style="font-size: 12px;"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Max
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root css-qtsj9i-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiFormLabel-colorPrimary css-7dax6m-MuiFormLabel-root"
          id="y-scale-group-label"
        >
          Scale
        </label>
        <div
          aria-labelledby="y-scale-group-label"
          class="MuiFormGroup-root MuiFormGroup-row css-qfz70r-MuiFormGroup-root"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input css-1m9pwf3"
                name="y scale radio buttons group"
                type="radio"
                value="linear"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11zohuh-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              Linear
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                name="y scale radio buttons group"
                type="radio"
                value="logarithmic"
              />
              <span
                class="css-hyxlzm"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hbvpl3-MuiSvgIcon-root"
                  data-testid="RadioButtonUncheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hhw7if-MuiSvgIcon-root"
                  data-testid="RadioButtonCheckedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              Log
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1pddimr-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
        >
          Displayed table channels
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-i06ziq-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            role="button"
            tabindex="0"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            data-testid="select displayed table channels"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Displayed table channels
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item css-1f064cs-MuiGrid-root"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon css-qm9mjt-MuiAutocomplete-root"
        role="autocomplete"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="select data channels"
            id="select data channels-label"
            style="font-size: 12px;"
          >
            Search all channels
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gywuxd-MuiInputBase-root-MuiOutlinedInput-root"
            style="font-size: 12px;"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-12yjm75-MuiInputBase-input-MuiOutlinedInput-input"
              id="select data channels"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                  data-testid="CloseIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1in441m"
              >
                <span>
                  Search all channels
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
